### PR TITLE
SSE-2636: Store and retrieve the Auth API key securely

### PR DESCRIPTION
**Pull request by abdul-gds**
# Code Changes
- add Kms Key Arn  in auth lambda
- add Kms Key for auth api key
- add Auth Kms Key Alias

### CODE REVIEW

File: infrastructure/api/sse-api.yml [hunk 1]
```diff
@@ -317,10 +318,47 @@ Resources:
...
+            Resource: '*'
+      KeySpec: ECC_NIST_P256
+      KeyUsage: ENCRYPT_DECRYPT
+
+  AuthKmsKeyAlias:
+    Type: AWS::KMS::Alias
+    Properties:
+      AliasName: !Sub "sse_auth_kms_Key"
+      TargetKeyId: !Ref AuthKmsKey
+
```
Comment (CharlesIC) [inline infrastructure/api/sse-api.yml:361]: It doesn't look like this is used anywhere?

File: infrastructure/api/sse-api.yml [hunk 2]
```diff
@@ -317,10 +318,47 @@ Resources:
...
+        - Statement:
+            - Effect: Allow
+              Action:
+                - "kms:Decrypt"
+                - "kms:Encrypt"
+              Resource:
+                - !Ref AuthKmsKey
+  
+ # KMS configuration for auth kms key
+  AuthKmsKey:
```
Comment (CharlesIC) [inline infrastructure/api/sse-api.yml:336]:
```suggestion
 EnvironmentEncryptionKey:
```
This is used by our Lambdas to encrypt their environment, this name would probably be clearer

**Review by CharlesIC — COMMENTED** [review]
The PR looks good, just a few small suggestions and the commits need to be squashed

File: infrastructure/api/sse-api.yml [hunk 3]
```diff
@@ -317,10 +318,47 @@ Resources:
...
+    Type: AWS::KMS::Key
+    DeletionPolicy: Retain
+    UpdateReplacePolicy: Retain
```
[the remark this answers is not in the document]
Comment (abdul-gds) [reply infrastructure/api/sse-api.yml]: updated to use Delete

File: infrastructure/api/sse-api.yml (hunk 2 above)
Comment (abdul-gds) [reply infrastructure/api/sse-api.yml]: fixed

File: infrastructure/api/sse-api.yml (hunk 1 above)
Comment (abdul-gds) [reply infrastructure/api/sse-api.yml]: It is used as best practice and make it  easily to identify the key in the AWS KMS console.

File: infrastructure/api/sse-api.yml [hunk 4]
```diff
@@ -317,10 +318,47 @@ Resources:
...
+      Enabled: TRUE
+      KeyPolicy:
+        Version: '2012-10-17'
+        Statement:
+          - Sid: Enable IAM User Permissions
+            Effect: Allow
+            Principal:
+              AWS: !Sub arn:aws:iam::${AWS::AccountId}:root
+            Action: kms:*
+            Resource: '*'
```
Comment (CharlesIC) [inline infrastructure/api/sse-api.yml:352]: ✅ This is the default key policy
https://docs.aws.amazon.com/kms/latest/developerguide/key-policy-default.html

File: infrastructure/api/sse-api.yml [hunk 5]
```diff
@@ -317,10 +318,47 @@ Resources:
...
+      KeySpec: ECC_NIST_P256
+      KeyUsage: ENCRYPT_DECRYPT
+
+  AuthKmsKeyAlias:
+    Type: AWS::KMS::Alias
+    Properties:
+      AliasName: !Sub "sse_auth_kms_Key"
+      TargetKeyId: !Ref EnvironmentEncryptionKey
```
Comment (CharlesIC) [inline infrastructure/api/sse-api.yml:360]:
```suggestion
  EnvironmentEncryptionKeyAlias:
    Type: AWS::KMS::Alias
    Properties:
      AliasName: api-lambda-environment-encryption-key
      TargetKeyId: !Ref EnvironmentEncryptionKey
```
We'll need to use this key for other lambdas to encrypt their environments in the future, so let's separate it from Auth.

File: infrastructure/api/sse-api.yml [hunk 6]
```diff
@@ -317,10 +318,47 @@ Resources:
...
+      Policies:
+        - Statement:
+            - Effect: Allow
+              Action:
+                - "kms:Decrypt"
+                - "kms:Encrypt"
+              Resource:
+                - !Ref EnvironmentEncryptionKey
+  
+ # KMS configuration for auth kms key
```
Comment (CharlesIC) [inline infrastructure/api/sse-api.yml:335]:
```suggestion
```

File: infrastructure/api/sse-api.yml [hunk 7]
```diff
@@ -317,10 +318,47 @@ Resources:
...
+  EnvironmentEncryptionKey:
+    Type: AWS::KMS::Key
+    DeletionPolicy: Delete
+    UpdateReplacePolicy: Delete
+    # checkov:skip=CKV_AWS_7: "Ensure rotation for customer created CMKs is enabled"
+    Properties:
+      Description: Key for auth key encrypt/decrypt
```
Comment (CharlesIC) [inline infrastructure/api/sse-api.yml:342]:
```suggestion
      Description: Lambda environment variables encryption key
```

File: infrastructure/api/sse-api.yml (hunk 5 above)
Comment (abdul-gds) [reply infrastructure/api/sse-api.yml]: Alias name must start with `alias` i.e
`alias/api-lambda-environment-encryption-key`

**Review by CharlesIC — APPROVED** [review]

**Pull request merged**